##  HOUSE ROCKET
    Property buying and selling company where the CEO of the company would like to maximize the company's profit by finding good deals.
    
    Empresa de compra e venda de propriedades onde o CEO da empresa gostaria de maximizar o lucro da empresa encontrando bons negocios.

## New CEO's questions
    1. How many properties per level the dataset has ?
        - level_0: Price between 0.00 and  321.950;   
        - level_1: Price between 321.950 e  450.000;   
        - lavel_2: Price between 450.000 e 645.000;   
        - level_3: Price Above  645.000.
        
    2. Add the following information to properties:
          - Street name;   
          - house number;   
          - neighborhood;   
          - City name;   
          - State name;   
          - Place ID;
          - OSM Type: Open Street Map type;
          - Country:;
          - Country Code.
          
     3. Add the property's level on the map with a color and change the point's size on the map base on the price
     4. Add filter options for anyone doing your analysis
     
           - Visualize properties with waterfront or not;   
           - Filter properties by price;
           - Minimum living room size;           
           - Minimum numbers of bathrooms;
           - Max basement size;
           - Filter properties conditions;
           - Filter by year of construction.
           
     5. Add filter's options on the last dashboard 
     
           - Visualize values from a certain date;          
           - Filter by year of renovation;
           - Filter properties with waterfront or not.

     6. What's the average house price by "level" ?
     7. What's the average living room size of the properties by "size":
        
        - size_0 -> size between 0 and 1427 sqft;
        - size_1 -> size between 1427 and 1910 sqft;
        - size_2 -> size between 1910 and 2550 sqft;
        - size_3 -> size above 2550 sqft.



## As novas perguntas dos CEO
    
    1. Qual a quantidade de imóveis por nível?
        - Nivel_0: Preço entre U$ 0.00 e U$ 321.950;
        - Nivel_1: Preço entre U$ 321.950 e U$ 450.000;
        - Nivel_2: Preço entre U$ 450.000 e U$ 645.000;
        - Nivel_3: Preço acima U$ de 645.000;
        
    2. Adicione as seguintes informações ao imóvel:
        - O nome da Rua
        - O número do imóvel
        - O nome do Bairro
        - O nome da Cidade
        - O nome da Estado
        - Place ID: ldentificação da localização
        - OSM Type: Open Street Map type
        - Country: Nome do País
        - Country Code: Código do País
    3. Adicione o nível do imóvel no mapa como uma cor e o preço do imóvel como o tamanho do ponto no mapa.
    4. Adicione opções de filtros para eu fazer minhas próprias análises:
        - Eu quero escolher visualizar imóveis com vista para agua ou não
        - Eu quero escolher filtrar os imóveis até um certo valor de preço.
        - Tamanho mínimo da área da sala de estar.
        - Número mínimo de banheiros.
        - Tamanho máximo da área do porão.
        - Fltro das Condições do Imóvel.
        - Filtro por Ano de Construção.
    5. Adicionar opções de filtros no último dashboard enviado:
        - Eu quero visualizar somente valores a partir de uma data disponível para comprar.
        - Filtro por data disponível para compra.
        - Filtro por ano de renovação.
        - Filtro se possui vista para a água ou não.
    6. Qual a média do preço de compra dos imóveis por "Nível"?
        - level_0 -> Preço entre R$ 0 e R$ 321.9500
        - level_1 -> Preço entre R$ 321.950 e R$ 450.000
        - level_2 -> Preço entre R$ 450.000 e RS 645.000
    - level_3 -> Acima de R$ 645.000
    7. Qual a média do tamanho da sala de estar dos imóveis por "Size"?
        - Size_0 -> Tamanho entre 0 e 1427 sqft
        - Size_1 -> Tamanho entre 1427e 1910 sqft
        - Size_2 -> Tamanho entre 1910e 2550 sqft
        - Size_3 -> Tamanho acima de 2550 sqft   


## Planning the solution
    1. Planning the final product
        1.1 What i will deliver ?
            A: Email + 3 annexes.
        1.2 How the delivery will be ?
            A: - Email: text with questions and answers;
               - Annex: one report .csv file, one dashboard and a map, both with filters
    2. Process planning
        2.1 Were is the information ?
            A: dataset on kaggle
        2.2 How colect data ?
            A: public available
        2.3 Answer the questions
    3. Tool planning ?
            A: - Python;
               - Plotly;
               - Matplotlib;
               - Seaborn;
               - Jupyter notebook.

## Planejameno da solução
    1. Planejamento do produto final
        1.1 Oque eu vou entregar?
            R: - Email + 3 anexos:
               - Email: Perguntras | respostas
               - Anexos: um arquivo .csv com as novas informações requisitados, um dashboard com filtros e um mapa com filtros        
        1.2 Como vai ser a entrega?
            R: Perguntas | Respostas
    2. Planejamento do processo
        2.1 Onde está a informação?
            R: dataset no kaggle
        2.2 Como coletar as informações?
            R: diponiveis publicamente
        2.3 responder as perguntas?
    3. Planejamento do ferramentas?
            R:- Python;
               - Plotly;
               - Matplotlib;
               - Seaborn;
               - Jupyter notebook.

![purple-divider](https://user-images.githubusercontent.com/7065401/52071927-c1cd7100-2562-11e9-908a-dde91ba14e59.png)
## Imports

In [130]:
import pandas as pd 
import numpy as np
import seaborn as sns
import locale
import matplotlib.pyplot as plt
import plotly.express as px
from matplotlib import gridspec
from IPython.core.display import HTML
from set_currency import set_currency

In [131]:
def jupyter_settings ( ) :
    %matplotlib inline
    %pylab inline
    plt.style.use( 'bmh' )
    plt.rcParams ['figure.figsize'] = [25, 12]
    plt.rcParams ['font.size'] = 24
    display(HTML('<style>.container {width:100% !important; }</style>'))
    pd.options.display.max_columns = None
    pd.options.display.max_rows= None
    pd.set_option( 'display.expand_frame_repr', False )
    sns.set()
                 
jupyter_settings ();

Populating the interactive namespace from numpy and matplotlib


C:\Users\Lavin\anaconda3\lib\site-packages\IPython\core\magics\pylab.py:159: UserWarning:

pylab import has clobbered these variables: ['style']
`%matplotlib` prevents importing * from pylab and numpy



![purple-divider](https://user-images.githubusercontent.com/7065401/52071927-c1cd7100-2562-11e9-908a-dde91ba14e59.png)
## Read the file

In [132]:
df = pd.read_csv('data/kc_house_data_mod.csv',parse_dates=[1])

In [133]:
df.head()

id       date     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement    yr_built yr_renovated  zipcode      lat     long       standard  house_age dormitory_type condition_type
0  7129300520 2014-10-13  221900.0         3       1.00         1180      5650     1.0           0     0          3      7        1180              0  1955-01-01   1970-01-01    98178  47.5112 -122.257   low_standard  new_house          house        regular
1  6414100192 2014-12-09  538000.0         3       2.25         2570      7242     2.0           0     0          3      7        2170            400  1951-01-01   1991-01-01    98125  47.7210 -122.319   low_standard  new_house          house        regular
2  5631500400 2015-02-25  180000.0         2       1.00          770     10000     1.0           0     0          3      6         770              0  1933-01-01   1970-01-01    98028  47.7379 -122.233   low_standard  new_house      apartment        regular
3  2487200875 2014-12-09  604000.0         4       3.00         1960      5000     1.0           0     0          5      7        1050            910  1965-01-01   1970-01-01    98136  47.5208 -122.393  high_standard  new_house          house           good
4  1954400510 2015-02-18  510000.0         3       2.00         1680      8080     1.0           0     0          3      8        1680              0  1987-01-01   1970-01-01    98074  47.6168 -122.045   low_standard  new_house          house        regular

In [134]:
df.tail()

id       date     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement    yr_built yr_renovated  zipcode      lat     long      standard  house_age dormitory_type condition_type
21608   263000018 2014-05-21  360000.0         3       2.50         1530      1131     3.0           0     0          3      8        1530              0  2009-01-01   1970-01-01    98103  47.6993 -122.346  low_standard  new_house          house        regular
21609  6600060120 2015-02-23  400000.0         4       2.50         2310      5813     2.0           0     0          3      8        2310              0  2014-01-01   1970-01-01    98146  47.5107 -122.362  low_standard  new_house          house        regular
21610  1523300141 2014-06-23  402101.0         2       0.75         1020      1350     2.0           0     0          3      7        1020              0  2009-01-01   1970-01-01    98144  47.5944 -122.299  low_standard  new_house      apartment        regular
21611   291310100 2015-01-16  400000.0         3       2.50         1600      2388     2.0           0     0          3      8        1600              0  2004-01-01   1970-01-01    98027  47.5345 -122.069  low_standard  new_house          house        regular
21612  1523300157 2014-10-15  325000.0         2       0.75         1020      1076     2.0           0     0          3      7        1020              0  2008-01-01   1970-01-01    98144  47.5941 -122.299  low_standard  new_house      apartment        regular

In [135]:
df.shape

(21613, 23)

### Check for duplicates

In [136]:
df.duplicated().value_counts()

False    21613
dtype: int64

In [137]:
df.duplicated(subset=['id']).value_counts()

False    21436
True       177
dtype: int64

The dataset has some duplicate ids, but the other columns are different, so if we check for duplicates with "pd.duplicated()", they will only be shown if we add the parameter "subset" with the "id" column as a reference

In [138]:
df.drop_duplicates(inplace=True)

In [139]:
df.isnull().sum()

id                 0
date               0
price              0
bedrooms           0
bathrooms          0
sqft_living        0
sqft_lot           0
floors             0
waterfront         0
view               0
condition          0
grade              0
sqft_above         0
sqft_basement      0
yr_built           0
yr_renovated       0
zipcode            0
lat                0
long               0
standard           0
house_age          0
dormitory_type    13
condition_type     0
dtype: int64

### Checking columns types

In [140]:
df.dtypes

id                         int64
date              datetime64[ns]
price                    float64
bedrooms                   int64
bathrooms                float64
sqft_living                int64
sqft_lot                   int64
floors                   float64
waterfront                 int64
view                       int64
condition                  int64
grade                      int64
sqft_above                 int64
sqft_basement              int64
yr_built                  object
yr_renovated              object
zipcode                    int64
lat                      float64
long                     float64
standard                  object
house_age                 object
dormitory_type            object
condition_type            object
dtype: object

![purple-divider](https://user-images.githubusercontent.com/7065401/52071927-c1cd7100-2562-11e9-908a-dde91ba14e59.png)
### 1. How many properties per level the dataset has ?
    
   **- level_0: Price between 0.00 and  321.950**
   
   **- level_1: Price between 321.950 e  450.000***
   
   **- lavel_2: Price between 450.000 e 645.000***
   
   **- level_3: Price Above  645.000***
 

In [141]:
df['level'] = 'NA'

In [142]:
df.loc[(df['price']>=0)&(df['price']<321950),'level'] = 'level_0'
df.loc[(df['price']>=321950)&(df['price']<450000),'level'] = 'level_1'
df.loc[(df['price']>=450000)&(df['price']<645000),'level'] = 'level_2'
df.loc[(df['price']>=645000),'level'] = 'level_3'

In [143]:
df['level'].value_counts()

level_2    5508
level_3    5413
level_0    5403
level_1    5289
Name: level, dtype: int64

The dataset have 5508 houses in level_0, 5413 houses in level_01, 5403 houses in level_2 and 5289 houses in level_03

### 2. Add the following information to properties:
    
  **- Street name**
   
  **- house number**
   
 **- neighborhood**
   
  **- City name**
   
  **- State name**
   
**- Place ID**

**- OSM Type: Open Street Map type**

**- Country:**

**- Country Code**
   
   

In [144]:
from geopy.geocoders import Nominatim

In [145]:
df.head()

id       date     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement    yr_built yr_renovated  zipcode      lat     long       standard  house_age dormitory_type condition_type    level
0  7129300520 2014-10-13  221900.0         3       1.00         1180      5650     1.0           0     0          3      7        1180              0  1955-01-01   1970-01-01    98178  47.5112 -122.257   low_standard  new_house          house        regular  level_0
1  6414100192 2014-12-09  538000.0         3       2.25         2570      7242     2.0           0     0          3      7        2170            400  1951-01-01   1991-01-01    98125  47.7210 -122.319   low_standard  new_house          house        regular  level_2
2  5631500400 2015-02-25  180000.0         2       1.00          770     10000     1.0           0     0          3      6         770              0  1933-01-01   1970-01-01    98028  47.7379 -122.233   low_standard  new_house      apartment        regular  level_0
3  2487200875 2014-12-09  604000.0         4       3.00         1960      5000     1.0           0     0          5      7        1050            910  1965-01-01   1970-01-01    98136  47.5208 -122.393  high_standard  new_house          house           good  level_2
4  1954400510 2015-02-18  510000.0         3       2.00         1680      8080     1.0           0     0          3      8        1680              0  1987-01-01   1970-01-01    98074  47.6168 -122.045   low_standard  new_house          house        regular  level_2

In [146]:
# dataset = pd.DataFrame()

# for i in range(len(df)):
#     aux = {}
#     lat_long = df.iloc[i]['lat':'long'].astype(str)
#     geolocator = Nominatim(user_agent="geoapiExercises")
#     location = geolocator.reverse(",".join(lat_long))
    
#     if 'road' in location.raw['address']:
#         aux['street'] = location.raw['address']['road']
#     else:
#         aux['street'] = 'NA'
  
#     if 'house_number' in location.raw['address']:
#         aux['house_number'] = location.raw['address']['house_number']
#     else:
#         aux['house_number'] = 'NA'
        
#     if 'neighbourhood' in location.raw['address']:
#         aux['neighbourhood'] = location.raw['address']['neighbourhood']
#     else:
#         aux['neighbourhood'] = 'NA'
                      
#     if 'town' in location.raw['address']:
#         aux['city'] = location.raw['address']['town']
#     elif 'city' in location.raw['address']:
#         aux['city'] = location.raw['address']['city']
#     else:
#         aux['city'] = 'NA'
                
#     if 'state' in location.raw['address']:
#         aux['state'] = location.raw['address']['state']
#     else:
#         aux['state'] = 'NA'
        
#     if 'place_id' in location.raw:
#         aux['place_id'] = location.raw['place_id']
#     else:
#         aux['place_id'] = 'NA'
        
#     if 'osm_type' in location.raw:
#         aux['osm_type'] = location.raw['osm_type']
#     else:
#         aux['osm_type'] = 'NA'  
        
#     if 'country' in location.raw['address']:
#         aux['country'] = location.raw['address']['country']
#     else:
#         aux['country'] = 'NA'
                
#     if 'country_code' in location.raw['address']:
#         aux['country_code'] = location.raw['address']['country_code']
#     else:
#         aux['country_code'] = 'NA'
              
        
#     data = pd.DataFrame(aux,index=[0])
#     dataset = pd.concat([dataset,data],axis=0)
#     dataset.to_csv('data/address_complet.csv',index=False)

Request with geopy passing "lat" and "long" will return the address information, and after that we save it to "address_complet.csv".

**This sections is commented, because demand a lot if time to complete, so to avoid accidentally run it again we keep commented**

In [147]:
df_address = pd.read_csv('data/address_complet.csv')

In [148]:
df_address.shape

(21613, 10)

In [149]:
df_address.columns

Index(['Unnamed: 0', 'street', 'house_number', 'neighbourhood', 'city',
       'state', 'place_id', 'osm_type', 'country', 'country_code'],
      dtype='object')

In [150]:
df_address.drop("Unnamed: 0",axis=1,inplace=True)

In [151]:
df_address.shape[0] == df.shape[0]

True

Check if both dataframes have the same number of rows

In [152]:
df = pd.concat([df,df_address],axis=1)

Join then

In [153]:
df.head()

id       date     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement    yr_built yr_renovated  zipcode      lat     long       standard  house_age dormitory_type condition_type    level                    street house_number  neighbourhood       city       state   place_id osm_type        country country_code
0  7129300520 2014-10-13  221900.0         3       1.00         1180      5650     1.0           0     0          3      7        1180              0  1955-01-01   1970-01-01    98178  47.5112 -122.257   low_standard  new_house          house        regular  level_0         61st Avenue South        10012  Rainier Beach    Seattle  Washington  161022547      way  United States           us
1  6414100192 2014-12-09  538000.0         3       2.25         2570      7242     2.0           0     0          3      7        2170            400  1951-01-01   1991-01-01    98125  47.7210 -122.319   low_standard  new_house          house        regular  level_2    Northeast 127th Street          837            NaN    Seattle  Washington  160016678      way  United States           us
2  5631500400 2015-02-25  180000.0         2       1.00          770     10000     1.0           0     0          3      6         770              0  1933-01-01   1970-01-01    98028  47.7379 -122.233   low_standard  new_house      apartment        regular  level_0     81st Avenue Northeast        15109            NaN    Kenmore  Washington   74951055     node  United States           us
3  2487200875 2014-12-09  604000.0         4       3.00         1960      5000     1.0           0     0          5      7        1050            910  1965-01-01   1970-01-01    98136  47.5208 -122.393  high_standard  new_house          house           good  level_2  Fauntleroy Way Southwest         9236            NaN    Seattle  Washington  156726151      way  United States           us
4  1954400510 2015-02-18  510000.0         3       2.00         1680      8080     1.0           0     0          3      8        1680              0  1987-01-01   1970-01-01    98074  47.6168 -122.045   low_standard  new_house          house        regular  level_2    221st Avenue Northeast          923            NaN  Sammamish  Washington  296045104      way  United States           us

### 3. Add the property's level on the map with a color and change the point's size on the map base on the price


In [154]:
data_mapa = df[['id','lat','long','price','level','waterfront']].copy()

In [155]:
fig = px.scatter_mapbox(data_mapa,lat="lat",lon="long",color='level',size='price',hover_name="id",hover_data=["price"],color_continuous_scale=px.colors.cyclical.IceFire,zoom=8,height=700)
fig.update_layout(mapbox_style="open-street-map")
fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
fig.show()

### 4. Add filter options for anyone doing your analysis
**- Visualize properties with waterfront or not;**

**- Filter properties by price;**

**- Minimum living room size;**

**- Minimum numbers of bathrooms;**

**- Max basement size;**

**- Filter properties conditions;**

**- Filter by year of construction.**



In [156]:
import ipywidgets as widgets
from ipywidgets import fixed

To make the filters with a graphic interface we will use the ipywidgets library

In [157]:
data_mapa = df[['id','lat','long','price','level','waterfront','bathrooms','sqft_living','condition_type','sqft_basement']].copy()
data_mapa['yr_buid'] = pd.to_datetime(df['yr_built']).dt.strftime('%Y-%m')
data_mapa['waterfront'] = data_mapa['waterfront'].apply(lambda x: 'Yes' if x == 1 else 'No');
data_mapa.head()

id      lat     long     price    level waterfront  bathrooms  sqft_living condition_type  sqft_basement  yr_buid
0  7129300520  47.5112 -122.257  221900.0  level_0         No       1.00         1180        regular              0  1955-01
1  6414100192  47.7210 -122.319  538000.0  level_2         No       2.25         2570        regular            400  1951-01
2  5631500400  47.7379 -122.233  180000.0  level_0         No       1.00          770        regular              0  1933-01
3  2487200875  47.5208 -122.393  604000.0  level_2         No       3.00         1960           good            910  1965-01
4  1954400510  47.6168 -122.045  510000.0  level_2         No       2.00         1680        regular              0  1987-01

In [158]:
data_mapa['yr_buid'].value_counts()

2014-01    559
2006-01    454
2005-01    450
2004-01    433
2003-01    422
2007-01    417
1977-01    417
1978-01    387
1968-01    381
2008-01    367
1967-01    350
1979-01    343
1959-01    334
1990-01    320
1962-01    312
2001-01    305
1954-01    305
1987-01    294
1989-01    290
1969-01    280
1955-01    271
1988-01    270
1999-01    265
1947-01    263
1963-01    256
1976-01    253
1966-01    250
1950-01    250
1994-01    249
1960-01    248
1980-01    240
1998-01    239
1948-01    235
2009-01    230
1984-01    229
1951-01    229
1985-01    228
1991-01    224
1961-01    224
1958-01    224
1942-01    223
1953-01    223
2002-01    222
1952-01    220
2000-01    218
1986-01    215
1983-01    212
1993-01    202
2013-01    201
1981-01    199
1992-01    198
1956-01    198
1957-01    198
1996-01    195
1949-01    195
1975-01    189
1965-01    187
1926-01    180
1997-01    177
1964-01    172
2012-01    170
1943-01    170
1995-01    169
1925-01    165
1974-01    162
1941-01    161
1940-01   

In [159]:
data_mapa['waterfront'].value_counts()

No     21450
Yes      163
Name: waterfront, dtype: int64

In [160]:
style = {'description_width':'initial'}
waterfront_bar = widgets.ToggleButtons(
    options=['No','Yes'],
    description='Waterfront',
    disabled=False,
    button_style='', 
    tooltips=['yes','no'],
    style = style

)
condition_bar = widgets.ToggleButtons(
    options=['bad','regular','good'],
    description='Condition  ',
    disabled=False,
    button_style='', 
    tooltips=['bad','regular','good'],
    style = style

)
price_limit = widgets.IntRangeSlider(
    value=[int(data_mapa['price'].min()), int(data_mapa['price'].max())],
    min = int(data_mapa['price'].min()),
    max = int(data_mapa['price'].max()),
    step=1,
    description='Price range',
    disabled=False,
    continuous_update=False,
    orientation='horizontal',
    readout=True,
    readout_format='d',
   style = style
)
bathroom_limit = widgets.IntSlider(
    value = 0,
    min = data_mapa['bathrooms'].min(),
    max = data_mapa['bathrooms'].max(),
    step = 1,
    description = 'Bathrooms',
    disable=False,
    style = style
)
liv_limit = widgets.IntRangeSlider(
    value=[data_mapa['sqft_living'].min(), data_mapa['sqft_living'].max()],
    min = data_mapa['sqft_living'].min(),
    max = data_mapa['sqft_living'].max(),
    step=10,
    description='Living size',
    disabled=False,
    continuous_update=False,
    orientation='horizontal',
    readout=True,
    readout_format='d',
    style = style
)
base_limit = widgets.IntRangeSlider(
    value=[data_mapa['sqft_basement'].min(), data_mapa['sqft_basement'].max()],
    min = data_mapa['sqft_basement'].min(),
    max = data_mapa['sqft_basement'].max(),
    step=10,
    description='Basement size',
    disabled=False,
    continuous_update=False,
    orientation='horizontal',
    readout=True,
    readout_format='d',
    style = style
)


def updata_map(df,cond, waterfront,limit,bath,liv,base):
    houses = df[((df['price'] >= limit[0])&(df['price'] <= limit[1])) & (df['waterfront'] == waterfront) & (df['bathrooms'] <= bath) & ((df['sqft_living']>=liv[0])& (df['sqft_living']<=liv[1]))&(df['condition_type']==cond) & ((df['sqft_basement']>=base[0])& (df['sqft_basement']<=base[1]))].copy()
    print("There are:",len(houses),"houses available with this conditions")
    fig = px.scatter_mapbox(houses,lat="lat",lon="long",color='level',size='price',hover_name="id",hover_data=["price","bathrooms","sqft_living","sqft_basement"],size_max=15,color_continuous_scale=px.colors.cyclical.IceFire,zoom=8,height=800)
    fig.update_layout(mapbox_style="open-street-map")
    fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
    fig.show()
    

In [161]:
widgets.interactive(updata_map,df = fixed(data_mapa),cond = condition_bar,waterfront = waterfront_bar,limit = price_limit,bath = bathroom_limit,liv=liv_limit,base = base_limit)

interactive(children=(ToggleButtons(description='Condition  ', options=('bad', 'regular', 'good'), style=Toggl…

### 5. Add filter's options on the last dashboard 
**- Visualize values from a certain date;** 

**- Filter by year of renovation;**

**- Filter properties with waterfront or not.**



In [162]:
df_dash = df.copy()

copying the original dataframe to be able to do some transformations without changing the data

In [163]:
df_dash['year'] = pd.to_datetime(df_dash['date']).dt.strftime('%y')
df_dash['date'] = pd.to_datetime(df_dash['date']).dt.strftime('%Y-%m-%d')
df_dash['yr_renovated'] = pd.to_datetime(df_dash['yr_renovated']).dt.year
df_dash['year_week'] = pd.to_datetime(df_dash['date']).dt.strftime('%Y-%U')

In [164]:
df_dash.head()

id        date     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement    yr_built  yr_renovated  zipcode      lat     long       standard  house_age dormitory_type condition_type    level                    street house_number  neighbourhood       city       state   place_id osm_type        country country_code year year_week
0  7129300520  2014-10-13  221900.0         3       1.00         1180      5650     1.0           0     0          3      7        1180              0  1955-01-01          1970    98178  47.5112 -122.257   low_standard  new_house          house        regular  level_0         61st Avenue South        10012  Rainier Beach    Seattle  Washington  161022547      way  United States           us   14   2014-41
1  6414100192  2014-12-09  538000.0         3       2.25         2570      7242     2.0           0     0          3      7        2170            400  1951-01-01          1991    98125  47.7210 -122.319   low_standard  new_house          house        regular  level_2    Northeast 127th Street          837            NaN    Seattle  Washington  160016678      way  United States           us   14   2014-49
2  5631500400  2015-02-25  180000.0         2       1.00          770     10000     1.0           0     0          3      6         770              0  1933-01-01          1970    98028  47.7379 -122.233   low_standard  new_house      apartment        regular  level_0     81st Avenue Northeast        15109            NaN    Kenmore  Washington   74951055     node  United States           us   15   2015-08
3  2487200875  2014-12-09  604000.0         4       3.00         1960      5000     1.0           0     0          5      7        1050            910  1965-01-01          1970    98136  47.5208 -122.393  high_standard  new_house          house           good  level_2  Fauntleroy Way Southwest         9236            NaN    Seattle  Washington  156726151      way  United States           us   14   2014-49
4  1954400510  2015-02-18  510000.0         3       2.00         1680      8080     1.0           0     0          3      8        1680              0  1987-01-01          1970    98074  47.6168 -122.045   low_standard  new_house          house        regular  level_2    221st Avenue Northeast          923            NaN  Sammamish  Washington  296045104      way  United States           us   15   2015-07

In [177]:
date_limit = widgets.SelectionSlider(
    options = df_dash['date'].sort_values().unique().tolist(),
    value = '2014-12-01',
    description = 'Disponivel',
    continuous_update = False,
    orientation = 'horizontal',
    readout= True
)
year_limit = widgets.SelectionSlider(
    options = df_dash['yr_renovated'].sort_values().unique().tolist(),
    value = 2000,
    description = 'Max year',
    continuous_update = False,
    orientation = 'horizontal',
    readout= True
)
waterfront_bar = widgets.ToggleButtons(
    options=['no','yes'],
    description='Waterfront',
    disabled=False,
    button_style='', 
    tooltips=['no','yes'],

)
def update_dash(data,limit,year,front):
    #filter data
    data['waterfront'] = data['waterfront'].apply(lambda x: 'yes' if x == 1 else 'no')
    df = data[(data['date'] <=limit)&(data['yr_renovated']>=year)&(data['waterfront']==front)].copy()
    
    fig = plt.figure(figsize=(21,12))
    specs= gridspec.GridSpec(ncols=2,nrows=2,figure = fig)
    
    ax1 = fig.add_subplot(specs[0,:])
    ax2 = fig.add_subplot(specs[1,0])
    ax3 = fig.add_subplot(specs[1,1])
    
    by_year = df[['price','year']].groupby('year').sum().reset_index()
    ax1.bar(by_year['year'],by_year['price'])
    set_currency(ax1)
    ax1.set_title('Avg Price by Year')
    
    by_day = df[['price','date']].groupby('date').mean().reset_index()
    ax2.plot(by_day['date'],by_day['price'])
    ax2.set_xticklabels(by_day['date'].unique().tolist(),rotation=60)
    set_currency(ax2)
    ax2.set_title('Avg Price by Day')

    
    by_year_week = df[['price','year_week']].groupby('year_week').mean().reset_index()
    ax3.bar(by_year_week['year_week'],by_year_week['price'])
    set_currency(ax3)
    ax3.set_title('Avg Price by Year Week')
    plt.xticks(rotation=60);


In [179]:
widgets.interact(update_dash,data = fixed(df_dash),limit = date_limit,year = year_limit,front = waterfront_bar);

interactive(children=(SelectionSlider(continuous_update=False, description='Disponivel', index=323, options=('…

### 6. What's the average house price by "level" ?


In [167]:
avg_price_level = df[['price','level']].groupby('level').mean().applymap(lambda x:locale.currency(x,grouping=True))
avg_price_level.reset_index()

level        price
0  level_0  $251,544.62
1  level_1  $383,585.21
2  level_2  $536,164.42
3  level_3  $985,008.98

###     7. What's the average living room size of the properties by "size"
**- size_0 -> size between 0 and 1427 sqft**

**- size_1 -> size between 1427 and 1910 sqft**

**- size_2 -> size between 1910 and 2550 sqft**

**- size_3 -> size above 2550 sqft**


In [168]:
df['size'] = 'NA'

In [169]:
df.loc[(df['sqft_living'] >= 0) & (df['sqft_living'] < 1427),'size'] = 'size_0'
df.loc[(df['sqft_living'] >= 1427) & (df['sqft_living'] < 1910),'size'] = 'size_1'
df.loc[(df['sqft_living'] >= 1910) & (df['sqft_living'] < 2550),'size'] = 'size_2'
df.loc[(df['sqft_living'] >= 2550),'size'] = 'size_3'

In [170]:
df['size'].value_counts()

size_2    5441
size_3    5433
size_0    5403
size_1    5336
Name: size, dtype: int64

In [171]:
avg_liv_size = df[['sqft_living','size']].groupby('size').mean().reset_index()
avg_liv_size

size  sqft_living
0  size_0  1123.776421
1  size_1  1661.198088
2  size_2  2202.700239
3  size_3  3318.988404

In [181]:
df.to_csv("data/kc_house_data_mod.csv")

Save the new columns